pkg/api/metrics.go

```diff
@@ -97,67 +97,63 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 	return toMacronResponse(resp)
 }
 
-func toMacronResponse(qdr *backend.QueryDataResponse) response.Response {
-	statusCode := http.StatusOK
-	for _, res := range qdr.Responses {
-		if res.Error != nil {
-			statusCode = http.StatusBadRequest
-		}
-	}
-
-	return response.JSONStreaming(statusCode, qdr)
-}
+func (hs *HTTPServer) getDataSourceFromQuery(c *models.ReqContext, query *simplejson.Json, history map[string]*models.DataSource) (*models.DataSource, response.Response) {
+	var err error
+	uid := query.Get("datasource").Get("uid").MustString()
 
-// handleExpressions handles POST /api/ds/query when there is an expression.
-func (hs *HTTPServer) handleExpressions(c *models.ReqContext, reqDTO dtos.MetricRequest) response.Response {
-	timeRange := plugins.NewDataTimeRange(reqDTO.From, reqDTO.To)
-	request := plugins.DataQuery{
-		TimeRange: &timeRange,
-		Debug:     reqDTO.Debug,
-		User:      c.SignedInUser,
-		Queries:   make([]plugins.DataSubQuery, 0, len(reqDTO.Queries)),
+	// before 8.3 special types could be sent as datasource (expr)
+	if uid == "" {
+		uid = query.Get("datasource").MustString()
 	}
 
-	for _, query := range reqDTO.Queries {
-		hs.log.Debug("Processing metrics query", "query", query)
-		name := query.Get("datasource").MustString("")
+	// check cache value
+	ds, ok := history[uid]
+	if ok {
+		return ds, nil
+	}
 
-		datasourceID, err := query.Get("datasourceId").Int64()
+	switch uid {
+	case expr.OldDatasourceUID: // "-100" :(
+		fallthrough
+	case expr.DatasourceUID:
+		return expr.DataSourceModel(), nil
```

Comment:
Could maybe hide the details in expr package by exposing a function, e.g. `expr.IsDatasource(<uid>)`?

```suggestion
	if expr.IsDataSource(uid) {
		return expr.DataSourceModel(), nil
	}
	
	switch uid {
```

Reply:
I am torn on how to approach this -- my sense is we should fix/migrate everything so that `expr.DatasourceUID` is the real value!

However, fixing that along with everything else in this PR seems like a big lift.

Reply:
After this is merged, I think we should audit expressions to make sure:
1. "-100" is handled by migrations and not with every check at runtime
2. datasource loading does not fall back to looking in the `map[string]interface{}`

If adding the function now is helpful, then :+1:

Reply:
> However, fixing that along with everything else in this PR seems like a big lift.

Agree. Guess my point was moving the checks of these different constant behind a function would allow us to just remove usage of `OldDatasourceUID` in that function when we've migrated to use only `DatasourceUID`. Guess it could wait